src/object/obj_class.c

```diff
@@ -230,16 +226,21 @@ daos_oclass_fit_max(daos_oclass_id_t oc_id, int domain_nr, int target_nr,
 	}
 
 	grp_size = daos_oclass_grp_size(&ca);
-	if (ca.ca_grp_nr == DAOS_OBJ_GRP_MAX ||
-	    ca.ca_grp_nr * grp_size > target_nr) {
-		/* search for the highest scalability in the allowed range */
+	if (ca.ca_grp_nr == DAOS_OBJ_GRP_MAX)
 		ca.ca_grp_nr = max(1, (target_nr / grp_size));
-		oc = oclass_scale2cl(&ca);
-		*ord = oc->oc_redun;
-		*nr = ca.ca_grp_nr;
-		oc_id = oc->oc_id;
-		goto out;
+
+	if (grp_size > domain_nr) {
+		D_ERROR("grp size (%u) (%u) is larger than domain nr (%u)\n",
+			grp_size, DAOS_OBJ_REPL_MAX, domain_nr);
+		return -DER_INVAL;
+	}
+
+	if (ca.ca_grp_nr * grp_size > target_nr) {
+		D_ERROR("grp_size (%u) x grp_nr (%u) is larger than targets (%u)\n",
+			grp_size, ca.ca_grp_nr, target_nr);
+		return -DER_INVAL;
```

Comment:
When 'ca_grp_nr == DAOS_OBJ_GRP_MAX', I think we need to call oclass_scale2cl() to get the highest available 'oc' instead of return error, right?

Reply:
Now codes have been supported any number of groups, when ca_grp_nr = DAOS_OBJ_GRP_MAX, ca.ca_grp_nr will be assigned max(1, (target_nr / grp_size));
That is ok, we don't need search highest available "oc".